---
title: "BMEG 310 Final Project"
author: "Justin Yoon, Zeena ..., Addin ..."
date: "2024-11-28"
output: html_document
---

https://github.com/jyoon1325/BMEG310FinalProject.git

#Import libraries:
```{r}
#all libraries from tutorials:

#Introduction to R Basics
library(graphics)
#install.packages("ggplot2")
library(ggplot2)
#install.packages("RColorBrewer")
library(RColorBrewer)
#install.packages("rmarkdown")
library(rmarkdown)

#Introduction to Machine Learning
#install.packages("devtools")
library(devtools)
#install.packages("remotes")
#remotes::install_github("vqv/ggbiplot")
library(ggbiplot)
library(dplyr)
#install.packages("readxl")
library(readxl)#read_excel
#install.packages("ISLR")
require(ISLR)
#install.packages("corrplot")
library(corrplot)
#install.packages("caret")
library(caret)
#install.packages("randomForest")
library(randomForest)

#Introduction to Survival Analysis
#BiocManager::install("TCGAbiolinks")
library(TCGAbiolinks)
#BiocManager::install("survival")
library(survival)
#BiocManager::install("survminer")
library(survminer)
library(SummarizedExperiment)

#Introduction to Mutation Analysis
# install.packages("pheatmap")
library(pheatmap)

#Introduction to Differential Analysis
# BiocManager::install("AnnotationDbi")
# BiocManager::install("org.Hs.eg.db")
# BiocManager::install("pathview")
# BiocManager::install("gage")
# BiocManager::install("gageData")
library(gridExtra)
library(ggrepel)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(pathview)
library(gage)
library(gageData)


#Additional useful libraries:
library(tidyr)
library(readr)
library(data.table)
library(stats)
library(cluster)
library(factoextra)
library(org.Hs.eg.db)
library(enrichplot)
library(DESeq2)  
#install.packages(magrittr)
library (dplyr)
library(TCGAbiolinks)
#BiocManager::install("edgeR")
library(edgeR)

```


#Import data:
```{r}
RNA.data <- read.csv("RNAseq_LIHC.csv")
patient.data <- read.table("data_clinical_patient.txt", header = TRUE, sep = "\t")
mutations.data <- read.table("data_mutations.txt", header = TRUE, sep = "\t")
```

##summaries of data
```{r}
#summary(RNA.data)
#summary(patient.data)
#summary(mutations.data)
#colnames(RNA.data)
#colnames(patient.data)
#colnames(mutations.data)
```


#filter data
```{r}
patient_df<-patient.data
patient_df$PATIENT_ID<-substr(patient_df$PATIENT_ID,9,12)

mutations_df<-mutations.data
mutations_df$Tumor_Sample_Barcode<-substr(mutations_df$Tumor_Sample_Barcode,9,12)

#make copy of RNA data frame without the first column
RNA_df <- RNA.data[,-1]

#rename columns to match patient id naming convention of other data sets
colnames(RNA_df)<-substr(colnames(RNA_df),9,12)

#find unique patients
unique_patients<-unique(patient_df$PATIENT_ID)
unique_mutations<-unique(mutations_df$Tumor_Sample_Barcode)
unique_RNA<-unique(colnames(RNA_df))
#find common patients
common_patients<-Reduce(intersect, list(unique_patients,unique_mutations,unique_RNA))

#patient data with patients in all three data sets
unq_patient_df<-patient_df[patient_df$PATIENT_ID %in% common_patients,]

#mutation data with patients in all three data sets
unq_mutations_df<-mutations_df[mutations_df$Tumor_Sample_Barcode %in% common_patients,]

#RNA data with patients in all three data sets
unq_RNA_ID<-RNA_df[, colnames(RNA_df)%in% common_patients]

#add back the first column (X) from the RNA data frame
RNA_df$X<-RNA.data$X[match(rownames(RNA_df), rownames(RNA.data))] 
 

#Take out Low impact mutations, intron and synonymous variants
filtered_mutations <- subset(unq_mutations_df,
                             ((unq_mutations_df$IMPACT != "LOW") & (unq_mutations_df$Consequence != "intron_variant") 
                              & (unq_mutations_df$Consequence != "synonymous_variant")))

```
##Summaries of filtered data
```{r}
#summary(unq_patient_df)
#summary(filtered_mutations)
#summary(RNA_df)

```


#Analysis
_______________________________________________________________________________________________________________________________________

Analysis:
1. Clinical Data
2. Mutation Data
3. Expression Data

_______________________________________________________________________________________________________________________________________




##Clinical Data:
_______________________________________________________________________________________________________________________________________
Clinical Data:
1. Explore clinical data wrt the distribution based on various variable/attributes such as age, stage, survival, etc 
_______________________________________________________________________________________________________________________________________

####Age Distribution
```{r}
ggplot(unq_patient_df, aes(x = AGE)) +
  geom_histogram(binwidth = 5, fill = "blue", color = "black", alpha = 0.7) +
  labs(title = "Age Distribution", x = "Age", y = "Count") +
  theme_minimal()
```
####cancer stage Distribution
```{r}
ggplot(unq_patient_df, aes(x = AGE, fill = OS_STATUS)) +
  geom_histogram(binwidth = 5, position = "dodge", alpha = 0.7) +
  labs(title = "Age Distribution by Overall Survival Status", x = "Age", y = "Count", fill = "Overall Survival Status") +
  theme_minimal()

```

####Age vs Overall Survival
```{r}
ggplot(unq_patient_df, aes(x = AGE, fill = OS_STATUS)) +
  geom_histogram(binwidth = 5, position = "dodge", alpha = 0.7) +
  labs(title = "Age Distribution by Overall Survival", x = "Age", y = "Count", fill = "Overall Survival") +
  theme_minimal()
```
####Age vs Progression-Free Survival (PFS) Status
```{r}
ggplot(unq_patient_df, aes(x = AGE, fill = PFS_STATUS)) +
  geom_histogram(binwidth = 5, position = "dodge", alpha = 0.7) +
  labs(title = "Age Distribution by Progression-Free Survival Status", x = "Age", y = "Count", fill = "Progression-Free Survival Status") +
  theme_minimal()
```
####Age Distribution with Boxplot Grouped by Tumor Stage
```{r}
ggplot(unq_patient_df, aes(x = AJCC_PATHOLOGIC_TUMOR_STAGE, y = AGE, fill = AJCC_PATHOLOGIC_TUMOR_STAGE)) +
  geom_boxplot(alpha = 0.7) +
  labs(title = "Age Distribution by Tumor Stage", x = "Tumor Stage", y = "Age") +
  theme_minimal() +
  theme(legend.position = "none")
```
####Survival Distribution by Age
```{r}
ggplot(unq_patient_df, aes(x = AGE, fill = DSS_STATUS)) +
  geom_histogram(binwidth = 5, position = "stack", alpha = 0.7) +
  labs(title = "Age Distribution by Disease-Specific Survival Status", x = "Age", y = "Count", fill = "Disease-Specific Survival Status") +
  theme_minimal()
```

##Survival analysis
###Prepare Data
```{r}
#survival data
unq_patient_df <- unq_patient_df %>%
  mutate(
    #convert OS_STATUS to a logical deceased indicator
    deceased = ifelse(OS_STATUS == "1:DECEASED", TRUE, FALSE),
    #convert OS_MONTHS to numeric
    overall_survival = as.numeric(OS_MONTHS)
  )
```

###Kaplan-Meier Analysis by Gender
```{r}
#fit survival model for gender
fit_gender <- survfit(Surv(overall_survival, deceased) ~ SEX, data = unq_patient_df)

#Kaplan-Meier plot for gender
ggsurvplot(fit_gender, data = unq_patient_df, pval = TRUE, risk.table = TRUE, risk.table.col = "strata", risk.table.height = 0.3)
```

###Kaplan-Meier Analysis by Weight
```{r}
# categorize patients by weight
unq_patient_df <- unq_patient_df %>%
  mutate(weight_group = case_when(
    WEIGHT < 50 ~ "Underweight",
    WEIGHT >= 50 & WEIGHT < 70 ~ "Normal weight",
    WEIGHT >= 70 & WEIGHT < 90 ~ "Overweight",
    WEIGHT >= 90 ~ "Obese",
    is.na(WEIGHT) ~ "Unknown"
  ))

#exclude patients with unknown weight
unq_patient_df_filtered <- unq_patient_df %>%
  filter(weight_group != "Unknown")

#fit survival model for weight groups
fit_weight <- survfit(Surv(overall_survival, deceased) ~ weight_group, data = unq_patient_df_filtered)

#Kaplan-Meier for weight groups
ggsurvplot(fit_weight, 
           data = unq_patient_df_filtered, 
           pval = TRUE,                        # Display p-value for log-rank test
           risk.table = TRUE,                  # Add risk table below the plot
           risk.table.col = "strata",          # Color risk table by strata
           risk.table.height = 0.3,            # Adjust risk table height
           title = "Kaplan-Meier Survival Analysis by Weight Group",
           xlab = "Time (Months)", 
           ylab = "Survival Probability")

```


###Kaplan-Meier Analysis by Tumor Stage
```{r}
# remove sub-stages of AJCC_PATHOLOGIC_TUMOR_STAGE
unq_patient_df <- unq_patient_df %>%
  mutate(tumor_stage = gsub("[abc]$", "", AJCC_PATHOLOGIC_TUMOR_STAGE))

#fit survival model by tumor stage
fit_stage <- survfit(Surv(overall_survival, deceased) ~ tumor_stage, data = unq_patient_df)

#Kaplan-Meier for tumor stage
ggsurvplot(fit_stage, 
           data = unq_patient_df, 
           pval = TRUE, 
           risk.table = TRUE, 
           risk.table.col = "strata", 
           risk.table.height = 0.3,
           title = "Survival by Tumor Stage")

```
###Kaplan-Meier Analysis by race
```{r}
#Group races 
  #"white", "asian", "black", and "others"...?
unq_patient_df <- unq_patient_df %>%
  mutate(race_group = ifelse(RACE %in% c("white", "asian", "black or african american"), RACE, "others"))

#fit survival model by race group
fit_race <- survfit(Surv(overall_survival, deceased) ~ race_group, data = unq_patient_df)

#Kaplan-Meier plot for race
ggsurvplot(fit_race, 
           data = unq_patient_df, 
           pval = TRUE, 
           risk.table = TRUE, 
           risk.table.col = "strata", 
           risk.table.height = 0.3,
           title = "Survival by Race Group")

```
###Kaplan-Meier Survival by Age Group
```{r}
# age groups
unq_patient_df <- unq_patient_df %>%
  mutate(age_group = case_when(
    AGE < 50 ~ "<50",
    AGE >= 50 & AGE < 65 ~ "50-65",
    AGE >= 65 ~ "65+"
  ))

#fit survival model by age group
fit_age <- survfit(Surv(overall_survival, deceased) ~ age_group, data = unq_patient_df)

#Kaplan-Meier
ggsurvplot(fit_age, 
           data = unq_patient_df, 
           pval = TRUE, 
           risk.table = TRUE, 
           risk.table.col = "strata", 
           risk.table.height = 0.3,
           title = "Survival by Age Group")

```









## Create Clusters Using Expression Data

### Pre-Processing 

```{r}
#Move Column "X" to the first position

RNA_df <- RNA_df[, c("X", setdiff(names(RNA_df), "X"))]

#remove column "X", as it is not a sample
preprocessed_RNA = RNA_df[,-1]

#convert to matrix
preprocessed_RNA = as.matrix(preprocessed_RNA) 

gene_names <- RNA_df$X  # Extract gene names
rownames(preprocessed_RNA) <- gene_names  # Set as row names

#filter data that has 0 or 1 read count across all samples

preprocessed_RNA <- preprocessed_RNA[rowSums(preprocessed_RNA)>1,]
```

```{r}
# CPM normalize data
cpm_RNA <- cpm(preprocessed_RNA)

#log normalize the counts
log_RNA <- log2(cpm_RNA + 1)

```

```{r}
#Find the 50 most variably expressed genes

#Find Variance for each gene
expression_variance <- apply(log_RNA, 1, var)

#Sort based on variance
sorted_expression <- sort(expression_variance, decreasing = TRUE)

#Get top 50 genes
top_genes <- names(sorted_expression)[1:50]

```

```{r}
#subset the expression data to only include the top 50 genes
top_RNA <- log_RNA[top_genes, ]

#Visualize using a heatmap
pheatmap(top_RNA, scale = "row", clustering_distance_rows = "euclidean", 
         clustering_distance_cols = "euclidean", main = "Top 50 Variable Genes")
```

### Clustering 

```{r}
# Create a distance matrix
# Transpose data to cluster
dist_matrix <- dist(t(top_RNA), method = "euclidean")

#Hierarchical clustering
hc <- hclust(dist_matrix, method = "ward.D2") #ward's method 

#plot
plot(hc, main = "Hierarchical Clustering of Samples", xlab = "", sub = "")
```
```{r}
#Cut the Dendrogram into Clusters
k <- 2
clusters <- as.data.frame(cutree(hc, k=k))
colnames(clusters) <- "Cluster"

table(clusters)
```

```{r}
#PCA 
transposed_RNA <- t(top_RNA)  #transpose
pca <- prcomp(transposed_RNA, scale = TRUE)

pca_df <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2], Cluster = factor(clusters$Cluster), PATIENT_ID = rownames(clusters))

#put in labels
pca_df$PATIENT_ID <- colnames(top_RNA)

#visualize with ggplot
ggplot(pca_df, aes(x = PC1, y = PC2, color = Cluster)) +
  geom_point(size =4) +
  labs(title = paste("PCA Plot with", k, "Clusters"), x = "PC1", y = "PC2") +
  theme_minimal()

```

```{r}
# Extract List of patients in each cluster

clust1_IDs <- names(clusters[clusters == 1])
clust2_IDs <- names(clusters[clusters == 2])

# Save patient names to lists
list_of_patients <- list(
  Cluster1 = clust1_IDs,
  Cluster2 = clust2_IDs
)
```



### Survival Analysis on Clinical Data to compare Patients in the Clusters




### Mutation ANalysis on Patients in Clusters to Find Mutational Patterns 

### DESeq2 Analysis on Expression Data to compare the clusters

- volcano plot to showcase the results of DESeq





