---
title: "BMEG 310 Final Project"
author: "Justin Yoon, Zeena ..., Addin ..."
date: "2024-11-28"
output: html_document
---

https://github.com/jyoon1325/BMEG310FinalProject.git

Import libraries:
```{r}
#all libraries from tutorials:

#Introduction to R Basics
library(graphics)
#install.packages("ggplot2")
library(ggplot2)
#install.packages("RColorBrewer")
library(RColorBrewer)
#install.packages("rmarkdown")
library(rmarkdown)

#Introduction to Machine Learning
#install.packages("devtools")
library(devtools)
#install.packages("remotes")
#remotes::install_github("vqv/ggbiplot")
library(ggbiplot)
suppressPackageStartupMessages(library(dplyr))
#install.packages("readxl")
library(readxl)#read_excel
#install.packages("ISLR")
require(ISLR)
#install.packages("corrplot")
library(corrplot)
#install.packages("caret")
library(caret)
#install.packages("randomForest")
library(randomForest)

#Introduction to Survival Analysis
#BiocManager::install("TCGAbiolinks")
library(TCGAbiolinks)
#BiocManager::install("survival")
library(survival)
#BiocManager::install("survminer")
library(survminer)
library(SummarizedExperiment)

#Introduction to Mutation Analysis
# install.packages("pheatmap")
library(pheatmap)

#Introduction to Differential Analysis
# BiocManager::install("AnnotationDbi")
# BiocManager::install("org.Hs.eg.db")
# BiocManager::install("pathview")
# BiocManager::install("gage")
# BiocManager::install("gageData")
library(gridExtra)
library(ggrepel)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(pathview)
library(gage)
library(gageData)


#Additional useful libraries:
library(tidyr)
library(readr)
library(data.table)
library(stats)
library(cluster)
library(factoextra)
library(org.Hs.eg.db)
library(enrichplot)
library(DESeq2)  
#install.packages(magrittr)
library (dplyr)
library(TCGAbiolinks)

```


Import data:
```{r}
RNA.data <- read.csv("RNAseq_LIHC.csv")
patient.data <- read.table("data_clinical_patient.txt", header = TRUE, sep = "\t")
mutations.data <- read.table("data_mutations.txt", header = TRUE, sep = "\t")
```

summaries of data
```{r}
#summary(RNA.data)
#summary(patient.data)
#summary(mutations.data)
#colnames(RNA.data)
#colnames(patient.data)
#colnames(mutations.data)
```


filter data
```{r}
patient_df<-patient.data
patient_df$PATIENT_ID<-substr(patient_df$PATIENT_ID,9,12)

mutations_df<-mutations.data
mutations_df$Tumor_Sample_Barcode<-substr(mutations_df$Tumor_Sample_Barcode,9,12)

#make copy of RNA data frame without the first column
RNA_df <- RNA.data[,-1]

#rename columns to match patient id naming convention of other data sets
colnames(RNA_df)<-substr(colnames(RNA_df),9,12)

#find unique patients
unique_patients<-unique(patient_df$PATIENT_ID)
unique_mutations<-unique(mutations_df$Tumor_Sample_Barcode)
unique_RNA<-unique(colnames(RNA_df))
#find common patients
common_patients<-Reduce(intersect, list(unique_patients,unique_mutations,unique_RNA))

#patient data with patients in all three data sets
unq_patient_df<-patient_df[patient_df$PATIENT_ID %in% common_patients,]

#mutation data with patients in all three data sets
unq_mutations_df<-mutations_df[mutations_df$Tumor_Sample_Barcode %in% common_patients,]

#RNA data with patients in all three data sets
unq_RNA_ID<-RNA_df[, colnames(RNA_df)%in% common_patients]

#add back the first column (X) from the RNA data frame
RNA_df$X<-RNA.data$X[match(rownames(RNA_df), rownames(RNA.data))] 

#Take out censored patients

censored_patients <-  unq_patient_df %>%
  filter(PFS_STATUS == "0:CENSORED") %>%
  pull(PATIENT_ID)

unq_patient_df<-unq_patient_df[!unq_patient_df$PATIENT_ID %in% censored_patients,]

unq_mutations_df <-  unq_mutations_df[!unq_mutations_df$Tumor_Sample_Barcode %in% censored_patients,]

RNA_df <- RNA_df[, !colnames(RNA_df) %in% censored_patients]


unq_mutations_df$IMPACT <- as.character(unq_mutations_df$IMPACT)

#Take out Low impact mutations
filtered_mutations <- subset(unq_mutations_df, IMPACT != "LOW")


unique(filtered_mutations$IMPACT)

```
Summaries
```{r}
#summary(unq_patient_df)
summary(filtered_mutations)
#summary(unq_RNA_ID)

```





_______________________________________________________________________________________________________________________________________

Analysis:
1. Clinical Data
2. Mutation Data
3. Expression Data

_______________________________________________________________________________________________________________________________________







_______________________________________________________________________________________________________________________________________
Clinical Data:
1. Explore clinical data wrt the distribution based on various variable/attributes such as age, stage, survival, etc 
_______________________________________________________________________________________________________________________________________

Age Distribution
```{r}
ggplot(unq_patient_df, aes(x = AGE)) +
  geom_histogram(binwidth = 5, fill = "blue", color = "black", alpha = 0.7) +
  labs(title = "Age Distribution", x = "Age", y = "Count") +
  theme_minimal()
```
cancer stage Distribution
```{r}
ggplot(unq_patient_df, aes(x = AGE, fill = OS_STATUS)) +
  geom_histogram(binwidth = 5, position = "dodge", alpha = 0.7) +
  labs(title = "Age Distribution by Overall Survival Status", x = "Age", y = "Count", fill = "Overall Survival Status") +
  theme_minimal()

```

Age vs Overall Survival
```{r}
ggplot(unq_patient_df, aes(x = AGE, fill = OS_STATUS)) +
  geom_histogram(binwidth = 5, position = "dodge", alpha = 0.7) +
  labs(title = "Age Distribution by Overall Survival", x = "Age", y = "Count", fill = "Overall Survival") +
  theme_minimal()
```
Age vs Progression-Free Survival (PFS) Status
```{r}
ggplot(unq_patient_df, aes(x = AGE, fill = PFS_STATUS)) +
  geom_histogram(binwidth = 5, position = "dodge", alpha = 0.7) +
  labs(title = "Age Distribution by Progression-Free Survival Status", x = "Age", y = "Count", fill = "Progression-Free Survival Status") +
  theme_minimal()
```
Age Distribution with Boxplot Grouped by Tumor Stage
```{r}
ggplot(unq_patient_df, aes(x = AJCC_PATHOLOGIC_TUMOR_STAGE, y = AGE, fill = AJCC_PATHOLOGIC_TUMOR_STAGE)) +
  geom_boxplot(alpha = 0.7) +
  labs(title = "Age Distribution by Tumor Stage", x = "Tumor Stage", y = "Age") +
  theme_minimal() +
  theme(legend.position = "none")
```
Survival Distribution by Age
```{r}
ggplot(unq_patient_df, aes(x = AGE, fill = DSS_STATUS)) +
  geom_histogram(binwidth = 5, position = "stack", alpha = 0.7) +
  labs(title = "Age Distribution by Disease-Specific Survival Status", x = "Age", y = "Count", fill = "Disease-Specific Survival Status") +
  theme_minimal()
```

Survival analysis
Prepare Data
```{r}
#survival data
unq_patient_df <- unq_patient_df %>%
  mutate(
    #convert OS_STATUS to a logical deceased indicator
    deceased = ifelse(OS_STATUS == "1:DECEASED", TRUE, FALSE),
    #convert OS_MONTHS to numeric
    overall_survival = as.numeric(OS_MONTHS)
  )
```

Kaplan-Meier Analysis by Gender
```{r}
#fit survival model for gender
fit_gender <- survfit(Surv(overall_survival, deceased) ~ SEX, data = unq_patient_df)

#Kaplan-Meier plot for gender
ggsurvplot(fit_gender, data = unq_patient_df, pval = TRUE, risk.table = TRUE, risk.table.col = "strata", risk.table.height = 0.3)
```

Kaplan-Meier Analysis by Weight
```{r}
# categorize patients by weight
unq_patient_df <- unq_patient_df %>%
  mutate(weight_group = case_when(
    WEIGHT < 50 ~ "Underweight",
    WEIGHT >= 50 & WEIGHT < 70 ~ "Normal weight",
    WEIGHT >= 70 & WEIGHT < 90 ~ "Overweight",
    WEIGHT >= 90 ~ "Obese",
    is.na(WEIGHT) ~ "Unknown"
  ))

#exclude patients with unknown weight
unq_patient_df_filtered <- unq_patient_df %>%
  filter(weight_group != "Unknown")

#fit survival model for weight groups
fit_weight <- survfit(Surv(overall_survival, deceased) ~ weight_group, data = unq_patient_df_filtered)

#Kaplan-Meier for weight groups
ggsurvplot(fit_weight, 
           data = unq_patient_df_filtered, 
           pval = TRUE,                        # Display p-value for log-rank test
           risk.table = TRUE,                  # Add risk table below the plot
           risk.table.col = "strata",          # Color risk table by strata
           risk.table.height = 0.3,            # Adjust risk table height
           title = "Kaplan-Meier Survival Analysis by Weight Group",
           xlab = "Time (Months)", 
           ylab = "Survival Probability")

```


Kaplan-Meier Analysis by Tumor Stage
```{r}
# remove sub-stages of AJCC_PATHOLOGIC_TUMOR_STAGE
unq_patient_df <- unq_patient_df %>%
  mutate(tumor_stage = gsub("[abc]$", "", AJCC_PATHOLOGIC_TUMOR_STAGE))

#fit survival model by tumor stage
fit_stage <- survfit(Surv(overall_survival, deceased) ~ tumor_stage, data = unq_patient_df)

#Kaplan-Meier for tumor stage
ggsurvplot(fit_stage, 
           data = unq_patient_df, 
           pval = TRUE, 
           risk.table = TRUE, 
           risk.table.col = "strata", 
           risk.table.height = 0.3,
           title = "Survival by Tumor Stage")

```
Kaplan-Meier Analysis by race
```{r}
#Group races 
  #"white", "asian", "black", and "others"...?
unq_patient_df <- unq_patient_df %>%
  mutate(race_group = ifelse(RACE %in% c("white", "asian", "black or african american"), RACE, "others"))

#fit survival model by race group
fit_race <- survfit(Surv(overall_survival, deceased) ~ race_group, data = unq_patient_df)

#Kaplan-Meier plot for race
ggsurvplot(fit_race, 
           data = unq_patient_df, 
           pval = TRUE, 
           risk.table = TRUE, 
           risk.table.col = "strata", 
           risk.table.height = 0.3,
           title = "Survival by Race Group")

```
Kaplan-Meier Survival Analysis by Age Group
```{r}
# age groups
unq_patient_df <- unq_patient_df %>%
  mutate(age_group = case_when(
    AGE < 50 ~ "<50",
    AGE >= 50 & AGE < 65 ~ "50-65",
    AGE >= 65 ~ "65+"
  ))

#fit survival model by age group
fit_age <- survfit(Surv(overall_survival, deceased) ~ age_group, data = unq_patient_df)

#Kaplan-Meier
ggsurvplot(fit_age, 
           data = unq_patient_df, 
           pval = TRUE, 
           risk.table = TRUE, 
           risk.table.col = "strata", 
           risk.table.height = 0.3,
           title = "Survival by Age Group")

```






_______________________________________________________________________________________________________________________________________
Mutation data:
1. Build a gene-patient matrix that contains mutation data. Within this matrix, "1" represents whether there is a mutation in a given gene and given patient and "0" represents no mutation. Note, when building this matrix, you can take into account the type of mutations (e.g., synonymous, non-synonymous, etc) 
2. Pick the top 20 or most frequently mutated genes from this matrix.  
3. Perform clustering on the resulting mutation matrix Perhaps try ward and ward.D2 linkage methods if you see narrow clusters.
4. Take the clusters of patients and do (a) DE analysis on all genes (not just top 500), (b) do pathway analysis on the DE genes that pass the adjusted p-value threshold (not just the top 20 genes) to see if specific pathways are hit, (c) see if outcome data between patient clusters is different, (d) see if any other clinical data is different when you compare patient clusters, (e) explore the correlation between the patient clusters that you have got from mutation and expression data - how much overlap do you see between the patient clusters?
_______________________________________________________________________________________________________________________________________

1. Build a gene-patient matrix that contains mutation data. Within this matrix, "1" represents whether there is a mutation in a given gene and given patient and "0" represents no mutation. Note, when building this matrix, you can take into account the type of mutations (e.g., synonymous, non-synonymous, etc)
```{r}
# Filter for specific mutation types
valid_mutation_types <- c("missense_variant", "nonsense_variant", "frameshift_variant")
filtered_mutations <- filtered_mutations[filtered_mutations$Consequence %in% valid_mutation_types, ]

# Ensure filtered_mutations is not empty
if (nrow(filtered_mutations) == 0) {
  stop("No mutations found matching the specified types.")
}

# Extract unique Gene (Hugo_Symbol) and Patient (Tumor_Sample_Barcode) pairs
gene_patient_data <- unique(filtered_mutations[, c("Hugo_Symbol", "Tumor_Sample_Barcode")])

# Add a binary column indicating mutation presence
gene_patient_data$mutated <- 1

# Create the gene-patient matrix
gene_patient_matrix <- reshape2::dcast(
  gene_patient_data,
  Hugo_Symbol ~ Tumor_Sample_Barcode,
  value.var = "mutated",
  fill = 0
)

# Check the resulting matrix
head(gene_patient_matrix)

```

2. Pick the top 20 or most frequently mutated genes from this matrix.  
```{r}
# Calculate the total number of mutations for each gene
gene_patient_numeric <- as.matrix(gene_patient_matrix[, -1])  # Exclude the Hugo_Symbol column
rownames(gene_patient_numeric) <- gene_patient_matrix$Hugo_Symbol  # Retain gene names

gene_mutation_counts <- rowSums(gene_patient_numeric)  # Sum mutations across patients

# Add mutation counts to the matrix
gene_patient_matrix$Mutation_Counts <- gene_mutation_counts

# Sort genes by mutation counts in descending order
sorted_genes <- gene_patient_matrix[order(-gene_patient_matrix$Mutation_Counts), ]

# Select the top 20 most frequently mutated genes
top_20_genes <- sorted_genes[1:20, ]

# View the top 20 genes
top_20_genes


```

3. Perform clustering on the resulting mutation matrix Perhaps try ward and ward.D2 linkage methods if you see narrow clusters.
```{r}

# Subset the mutation matrix for clustering (top 20 genes only)
top_20_matrix <- as.matrix(top_20_genes[, -c(1, ncol(top_20_genes))])  # Remove Hugo_Symbol and Mutation_Counts
rownames(top_20_matrix) <- top_20_genes$Hugo_Symbol

# Perform hierarchical clustering using "ward" and "ward.D2" linkage methods
dist_matrix <- dist(top_20_matrix, method = "binary")  # Binary distance for mutation data

# Clustering with Ward method
hc_ward <- hclust(dist_matrix, method = "ward.D")

# Clustering with Ward.D2 method
hc_ward_d2 <- hclust(dist_matrix, method = "ward.D2")

# Plot the dendrograms
par(mfrow = c(1, 2))  # Set up a side-by-side layout
plot(hc_ward, main = "Hierarchical Clustering (Ward)", xlab = "", sub = "", cex = 0.8)
plot(hc_ward_d2, main = "Hierarchical Clustering (Ward.D2)", xlab = "", sub = "", cex = 0.8)


```























```{r}
# Ensure consistent formatting of barcodes
filtered_mutations$Tumor_Sample_Barcode <- trimws(filtered_mutations$Tumor_Sample_Barcode)
names(clusters_ward) <- trimws(names(clusters_ward))

# Check for common barcodes
common_barcodes <- intersect(filtered_mutations$Tumor_Sample_Barcode, names(clusters_ward))


# Filter `filtered_mutations` to include only matching barcodes
filtered_mutations <- filtered_mutations[filtered_mutations$Tumor_Sample_Barcode %in% common_barcodes, ]
```
```{r}
# Assign clusters to patients in `filtered_mutations`
filtered_mutations$Cluster <- clusters_ward[match(filtered_mutations$Tumor_Sample_Barcode, names(clusters_ward))]

# Drop rows where Cluster is NA
filtered_mutations <- filtered_mutations[!is.na(filtered_mutations$Cluster), ]
dim(filtered_mutations)  # Confirm non-zero rows

```

```{r}
# Subset `gene_patient_matrix` to only include common patients
common_barcodes <- intersect(colnames(gene_patient_matrix)[-1], names(clusters_ward))
gene_patient_matrix <- gene_patient_matrix[, c("Hugo_Symbol", common_barcodes)]

# Ensure no NA values in the matrix
gene_patient_matrix[is.na(gene_patient_matrix)] <- 0

```





4a. Take the clusters of patients and do (a) DE analysis on all genes (not just top 500)

```{r}
# Check dimensions of the filtered_mutations dataset
dim(filtered_mutations)

# Check dimensions of the gene_patient_matrix
dim(gene_patient_matrix)

# Check length of Cluster assignments
length(clusters_ward)

# Ensure alignment of `Tumor_Sample_Barcode` with Cluster assignments
filtered_mutations$Cluster <- clusters_ward[match(filtered_mutations$Tumor_Sample_Barcode, names(clusters_ward))]

# Drop rows where Cluster is NA
filtered_mutations <- filtered_mutations[!is.na(filtered_mutations$Cluster), ]
dim(filtered_mutations)  # Confirm dimensions after filtering

```




```{r}
# Ensure filtered_mutations contains valid mutation data
filtered_mutations <- filtered_mutations[filtered_mutations$Consequence %in% c("missense_variant", "nonsense_variant", "frameshift_variant"), ]

# Ensure clusters align with patients
filtered_mutations$Cluster <- clusters_ward[match(filtered_mutations$Tumor_Sample_Barcode, names(clusters_ward))]

na.omit(filtered_mutations$Cluster)


# Ensure consistent barcodes and filter matrix
mutation_matrix <- table(filtered_mutations$Hugo_Symbol, filtered_mutations$Tumor_Sample_Barcode)
mutation_matrix <- as.data.frame.matrix(mutation_matrix)
mutation_matrix[mutation_matrix > 0] <- 1

# Add cluster information
mutation_matrix <- mutation_matrix[, common_barcodes]
mutation_matrix <- cbind(Cluster = clusters_ward[common_barcodes], mutation_matrix)

# Check dimensions and structure
dim(mutation_matrix)
head(mutation_matrix)

```









































_________________________________________________________________________________________________________________________________
Expression data:
1. Normalize expression data using DEseq2. This was explained and included in the lab document. 
2. Pick the top n (e.g., 500, 1000, ...) most variable genes. This could be done before step 1 or after step 1. 
3. Perform clustering on the resulting expression matrix. Perhaps try ward and ward.D2 linkage methods if you see narrow clusters.
4. Take the clusters of patients and do (a) DE analysis on all genes (not just top 500), (b) do pathway analysis on the DE genes that pass the adjusted p-value threshold (not just the top 20 genes) to see if specific pathways are hit, (c) see if outcome data between the patient clusters is different, (d) see if any other clinical data is different when you compare patient clusters, (e) see if patient clusters are enriched with any type of mutation data.   
_________________________________________________________________________________________________________________________________
```{r}

```


